nvtabular/inference/triton/data_conversions.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import itertools
+
+try:
+    import cudf
+    import cupy as cp
+except ImportError:
+    cudf = cp = None
+
+import numpy as np
+import pandas as pd
+
+from nvtabular.dispatch import _build_cudf_list_column
+from nvtabular.ops.operator import Supports
+
+
+def convert_format(tensors, kind, target_kind):
+    """Converts data from format 'kind' to one of the formats specified in 'target_kind'
+    This allows us to convert data to/from dataframe representations for operators that
+    only support certain reprentations
+    """
+
+    # this is all much more difficult because of multihot columns, which don't have
+    # great representations in dicts of cpu/gpu arrays. we're representing multihots
+    # as tuples of (values, offsets) tensors in this case - but have to do work at
+    # each step in terms of converting.
+    if kind & target_kind:
+        return tensors, kind
+
+    elif target_kind & Supports.GPU_DICT_ARRAY:
+        if kind == Supports.CPU_DICT_ARRAY:
+            return _convert_array(tensors, cp.array), Supports.GPU_DICT_ARRAY
+        elif kind == Supports.CPU_DATAFRAME:
+            return _pandas_to_array(tensors, False), Supports.GPU_DICT_ARRAY
+        elif kind == Supports.GPU_DATAFRAME:
+            return _cudf_to_array(tensors, False), Supports.GPU_DICT_ARRAY
+
+    elif target_kind & Supports.CPU_DICT_ARRAY:
```

Comment:
Does the order of these `elif` clauses imply that `CPU_DICT_ARRAY` is a higher performance choice than `GPU_DATAFRAME` (if available)?

Reply:
thats intentional - I'm finding that for most ops the CPU_DICT_ARRAY is substantially faster than the GPU_DATAFRAME (for instance normalize is about 60x faster) when run on just a couple of rows. This obviously depends on the batch size - and we may want to tune preferred representation on batch size later, but this is a good first pass (especially since we need to output to this format in triton anyways).

Reply:
Yeah, makes sense. I was mostly just curious! 😃